### PR TITLE
[Operator] index_select

### PR DESCRIPTION
**Pull request by GwokHiujin**
We have completed the development of the index_select operator. Specifically:

- The corresponding aTen operator is index_select
- Added accuracy test and perf test

Based on my previous findings, most libraries mimic torch and use the `gather` operator to indirectly achieve the functionality of `index_select`. Therefore, in this branch's implementation, I've also added a gather-based approach(See in the annotations). 

Below are the perf testing results on the Nvidia A100 platform, using both our triton kernel and the `torch.gather`-based implementation. It's **worth noting** that this branch **does not include our `gather` operator**. So if our `gather` operator is merged, causing calls to `torch.gather` to be replaced by it. In this situation, as we mentioned in our previous documentation, the perf of `index_select` will significantly decrease (clearly due to the bottleneck effect...).

---

*With `torch.gather`*:

```
test_reduction_perf.py Operator index_select Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.008192             0.01024
6144                   0.01536            0.070656
11264                  0.02048            0.016384
16384                   0.0256            0.017408
21504                  0.03072            0.090112
26624                 0.036864            0.022528
31744                 0.041984             0.09216
36864                 0.047104            0.091136
41984                 0.052224             0.03072
47104                 0.057344            0.033792
52224                 0.063488            0.095232
57344                 0.067584            0.038912
62464                 0.073728            0.041984
67584                 0.079872            0.044032
72704                 0.084992             0.04608
77824                 0.089088            0.098304
Operator index_select Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.008192             0.01024
6144                   0.01536            0.091136
11264                 0.021504            0.016384
16384                 0.026624            0.019456
21504                 0.031744             0.09216
26624                 0.037888            0.026624
31744                 0.044032            0.029696
36864                 0.049152            0.094208
41984                 0.053248            0.034816
47104                 0.059392            0.038912
52224                 0.065536            0.045056
57344                 0.069632            0.045056
62464                 0.075776            0.048128
67584                  0.08192            0.099328
72704                 0.088064            0.054272
77824                 0.093184            0.057344
Operator index_select Performance Test (torch.bfloat16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.009216            0.011264
6144                   0.01536            0.011264
11264                  0.02048             0.01536
16384                   0.0256            0.016384
21504                  0.03072             0.02048
26624                  0.03584            0.022528
31744                 0.041984              0.0256
36864                 0.048128            0.028672
41984                 0.052224             0.03072
47104                 0.057344            0.032768
52224                 0.063488            0.036864
57344                 0.067584            0.038912
62464                 0.073728             0.04096
67584                 0.079872            0.044032
72704                 0.084992            0.100352
77824                 0.090112            0.048128
```

*With our triton kernel*: 
```
test_reduction_perf.py Operator index_select Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.008192            0.227328
6144                   0.01536              0.1536
11264                  0.02048            0.172032
16384                 0.026624            0.248832
21504                  0.03072            0.323584
26624                 0.036864            0.417792
31744                 0.043008            0.548864
36864                 0.047104            0.666624
41984                 0.052224            0.826368
47104                 0.057344             0.99328
52224                 0.063488              1.1264
57344                 0.068608              1.3824
62464                 0.073728             1.57082
67584                 0.079872             1.70189
72704                 0.084992             1.91898
77824                 0.089088             2.17088
Operator index_select Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.009216            0.083968
6144                   0.01536            0.106496
11264                 0.021504            0.335872
16384                 0.026624            0.264192
21504                 0.031744            0.417792
26624                 0.037888            0.581632
31744                 0.044032            0.791552
36864                 0.049152            0.925696
41984                 0.053248             1.15507
47104                 0.059392               1.408
52224                 0.065536             1.45613
57344                 0.069632             2.01626
62464                   0.0768             2.13504
67584                  0.08192             2.29376
72704                 0.088064             2.43302
77824                 0.093184             2.75968
Operator index_select Performance Test (torch.bfloat16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.008192            0.192512
6144                   0.01536             0.09728
11264                  0.02048             0.20992
16384                   0.0256            0.449536
21504                  0.03072            0.324608
26624                 0.036864             0.41984
31744                 0.041984            0.550912
36864                 0.047104            0.667648
41984                 0.052224            0.828416
47104                 0.058368            0.994304
52224                 0.063488             1.10387
57344                 0.068608             1.38138
62464                 0.072704             1.56979
67584                 0.079872             1.70701
72704                 0.084992             1.91898
77824                 0.089088             2.16269
```

### CODE REVIEW

File: tests/test_reduction_ops.py [hunk 1]
```diff
@@ -670,3 +670,21 @@ def test_accuracy_vectornorm(shape, ord, dim, keepdim, dtype):
...
+@pytest.mark.parametrize("dim", DIM_LIST)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_index_select(shape, dim, dtype):
+    inp = torch.randn(shape, dtype=dtype, device="cuda")
+    index_size = inp.size(dim)
+    from math import floor
+
+    index = torch.randint(0, index_size, [floor(index_size * 0.8)], device="cuda")
+
+    ref_inp = to_reference(inp)
```
Comment (StrongSpoon) [inline tests/test_reduction_ops.py:685]: index is set on "cuda" device so it needs to_reference too.

File: src/flag_gems/ops/index_select.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    grid = lambda meta: (triton.cdiv(M, meta["BLOCK_M"]), triton.cdiv(N, meta["BLOCK_N"]))
+    index_select_kernel[grid](inp, out, M, N, index, index_len)
+    if dim != out.ndim - 1:
+        order = [i for i in range(out.ndim - 1)]
+        order.insert(dim, out.ndim - 1)
+        return out.permute(order)
+    else:
+        return out
+    """
```
Comment (StrongSpoon) [inline src/flag_gems/ops/index_select.py]: annotate by mistake?

File: src/flag_gems/ops/index_select.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    # with dim_compress
+    inp = dim_compress(inp, dim)
+    N = inp_shape[dim]
+    M = inp.numel() // N
+    index_len = index.numel()
+    out_shape = list(inp.shape)
+    out_shape[inp.ndim - 1] = index.numel()
+    out = torch.empty(out_shape, dtype=inp.dtype, device=inp.device)
+
+    grid = lambda meta: (triton.cdiv(M, meta["BLOCK_M"]), triton.cdiv(N, meta["BLOCK_N"]))
```
Comment (StrongSpoon) [inline src/flag_gems/ops/index_select.py]: maybe kernel parallel in M and index_len works better?

**Review by StrongSpoon — COMMENTED** [review]
remove torch.gather calling ;)

File: src/flag_gems/ops/index_select.py (hunk 2 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/index_select.py]: If `gather` can be optimized later, the gather version may outperform the current kernel version, so I am retaining the gather version in this commit. But it's ok to release the kernel version, as the gather version can be easily reimplemented if needed later :)

File: src/flag_gems/ops/index_select.py (hunk 3 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/index_select.py]: That sounds good! I modified the kernel for parallel in M and index_len, and the perf testing results below indicate a slight improvement(Though still behind torch).
```
test_reduction_perf.py Operator index_select Performance Test (torch.float16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.009216            0.188416
6144                   0.01536            0.191488
11264                  0.02048            0.165888
16384                   0.0256            0.248832
21504                  0.03072            0.323584
26624                  0.03584            0.417792
31744                 0.041984            0.550912
36864                 0.047104            0.668672
41984                 0.052224            0.826368
47104                 0.057344            0.994304
52224                 0.063488              1.1008
57344                 0.067584             1.38445
62464                 0.073728             1.56979
67584                 0.079872             1.70189
72704                 0.084992             1.91488
77824                 0.089088             2.15859
Operator index_select Performance Test (torch.float32)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.009216             0.28672
6144                   0.01536            0.282624
11264                 0.021504            0.294912
16384                 0.026624            0.330752
21504                 0.031744            0.439296
26624                 0.037888            0.577536
31744                 0.043008            0.790528
36864                 0.049152              0.9216
41984                 0.054272             1.15405
47104                 0.059392             1.40595
52224                 0.065536             1.45715
57344                 0.070656             2.00806
62464                 0.075776             2.13299
67584                  0.08192             2.29478
72704                  0.08704             2.43814
77824                  0.09216             2.75763
Operator index_select Performance Test (torch.bfloat16)
Size        Torch Latency (ms)   Gems Latency (ms)
--------------------------------------------------
1024                  0.008192            0.190464
6144                  0.014336            0.436224
11264                  0.02048            0.198656
16384                   0.0256            0.241664
21504                  0.03072            0.324608
26624                 0.036864             0.41984
31744                 0.041984            0.551936
36864                 0.047104            0.668672
41984                 0.052224            0.826368
47104                 0.057344            0.996352
52224                 0.063488             1.10182
57344                 0.068608              1.3865
62464                 0.073728             1.57082
67584                 0.078848             1.70291
72704                 0.084992             1.91181
77824                 0.090112             2.16678
```

File: tests/test_reduction_ops.py (hunk 1 above)
Comment (GwokHiujin) [reply tests/test_reduction_ops.py:685]: done

File: src/flag_gems/ops/index_select.py (hunk 2 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/index_select.py]: if gather performs better than index_select_kernel, I suggest straightly importing it from gems.

**Review by StrongSpoon — APPROVED** [review]
lgtm

File: src/flag_gems/ops/index_select.py (hunk 2 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/index_select.py]: Using `gather` from gems can significantly degrade performance, so let's maintain the kernel version until we can optimize `gather`.

**Pull request merged**